lib/compress/zstd_compress.c

```diff
@@ -2803,18 +2884,684 @@ static ZSTD_blockCompressor ZSTD_selectBlockCompressor(ZSTD_strategy strat, int
+                    curTotalMatchLength = curForwardMatchLength +
+                                          curBackwardMatchLength;
+
+                    if (curTotalMatchLength > bestMatchLength) {
+                        bestMatchLength = curTotalMatchLength;
+                        forwardMatchLength = curForwardMatchLength;
+                        backwardMatchLength = curBackwardMatchLength;
+                        bestEntry = cur;
+                    }
+                }
+            }
+        }
+
+        /* No match found -- continue searching */
+        if (bestEntry == NULL) {
+            ZSTD_ldm_makeEntryAndInsertByTag(hashTable, bucketOffsets,
+                                             rollingHash, hBits, current);
+            ip++;
+            continue;
+        }
+
+        /* Match found */
+        mLength = forwardMatchLength + backwardMatchLength;
+        ip -= backwardMatchLength;
+
+        /* Call the block compressor on the remaining literals */
+        {
+            U32 const matchIndex = bestEntry->offset;
+            const BYTE* const match = base + matchIndex - backwardMatchLength;
+            U32 const offset = (U32)(ip - match);
+
+            /* Overwrite rep codes */
+            for (i = 0; i < ZSTD_REP_NUM; i++)
+                seqStorePtr->rep[i] = repToConfirm[i];
+
+            /* Fill tables for block compressor */
+            ZSTD_ldm_limitedNextToUpdate(cctx, anchor);
+            ZSTD_ldm_fillTables(cctx, anchor);
+
+            /* Call block compressor and get remaining literals */
+            lastLiterals = blockCompressor(cctx, anchor, ip - anchor);
+            cctx->nextToUpdate = (U32)(ip - base);
+
+            /* Update repToConfirm with the new offset */
+            for (i = ZSTD_REP_NUM - 1; i > 0; i--)
+                repToConfirm[i] = repToConfirm[i-1];
+            repToConfirm[0] = offset;
+
+            /* Store the sequence with the leftover literals */
+            ZSTD_storeSeq(seqStorePtr, lastLiterals, ip - lastLiterals,
+                          offset + ZSTD_REP_MOVE, mLength - MINMATCH);
+        }
+
+        /* Insert the current entry into the hash table */
+        ZSTD_ldm_makeEntryAndInsertByTag(hashTable, bucketOffsets,
+                                         rollingHash, hBits,
+                                         (U32)(lastHashed - base));
+
+        /* Fill the hash table from lastHashed+1 to ip+mLength*/
+        assert(ip + backwardMatchLength == lastHashed);
+        rollingHash = ZSTD_ldm_fillLdmHashTable(hashTable, bucketOffsets,
+                                                rollingHash, lastHashed,
+                                                ip + mLength,
+                                                base, hBits);
+        lastHashed = ip + mLength - 1;
+        ip += mLength;
+        anchor = ip;
+
+        /* Check immediate repcode */
+        while ( (ip <= ilimit)
```

Comment:
I kept it because the compression ratio improved and it didn't seem to have any negative effects on speed. Although only very slightly (order of ~2-3% maybe). It could be removed if it's better for code clarity. Thoughts?